lua/lsp_extensions/dart/closing_labels.lua

```diff
@@ -0,0 +1,76 @@
+--[[
+## Closing Labels
+
+** Method: 'dart/textDocument/publishClosingLabels**
+Direction: Server -> Client Params: { uri: string, labels: { label: string, range: Range }[] }
+This notifies the client when closing label information is available (or updated) for a file.
+
+Since this is a notification, the callback needs to be registered in the client's callbacks table.
+This can be achieved with nvim_lspconfig with this minimal config.
+```lua
+nvim_lsp.dartls.setup{
+  init_options = {
+    closingLabels = true,
+  },
+  callbacks = {
+    ['dart/textDocument/publishClosingLabels'] = require('lsp_extensions.dart.closing_labels').get_callback{},
+  },
+}
+```
+
+https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/tool/lsp_spec/README.md#darttextdocumentpublishclosinglabels-notification
+--]]
+
+local M = {}
+-- Stored labels from the server's notifications keyed by uri
+local all_labels = {}
+-- Namespace for the virtual text
+local closing_labels_ns = vim.api.nvim_create_namespace('lsp_extensions.dart.closing_labels')
+-- Stored opts from the get_callback invocation.
+local label_opts = {}
+
+-- Gets a callback to register to the dartls publishClosingLabels notification.
+-- @tparam table a table of options: highlight, prefix
+M.get_callback = function(opts)
+  label_opts.highlight = opts.highlight or "Comment"
+  label_opts.prefix = opts.prefix or " // "
+
+  vim.cmd [[augroup DartShowClosingLabels]]
+  vim.cmd [[  au!]]
+  vim.cmd [[  autocmd CursorHold,CursorHoldI *.dart :lua require('lsp_extensions.dart.closing_labels').draw_labels()]]
+  vim.cmd [[augroup END]]
+
+  return function(_, _, result, _, _)
+    local uri = result.uri
+    local labels = result.labels
+    all_labels[uri] = labels
+    if uri == vim.uri_from_bufnr(0) then
+      M.draw_labels()
```

Comment:
you should make `draw_labels` take an `opts` parameter.

Pass the opts that you make above there, rather than saving it as an upvalue.

Reply:
I wanted to do that but I wanted to keep the autocommand as simple as possible. From your other comment maybe I should just ditch the auto command altogether and just document how the user could set it up themselves?

Reply:
Also, does it make sense that `draw_labels` and `get_callback` both take opts in that case? In order to get the same behavior after receiving the notification and on `CursorHold` you would need to pass the same opts. That felt wrong but maybe that is preferable to the approach I have here.

Reply:
I think what you have is a lot closer, but see my comment above.